src/main/java/ch/njol/skript/effects/EffPotion.java

```diff
@@ -36,143 +36,129 @@
 	@Override
-	public boolean init(final Expression<?>[] exprs, final int matchedPattern, final Kleenean isDelayed, final ParseResult parseResult) {
-		apply = matchedPattern > 0;
+	@SuppressWarnings("unchecked")
+	public boolean init(Expression<?>[] exprs, int matchedPattern, Kleenean isDelayed, ParseResult parseResult) {
 		potionEffect = matchedPattern == 0;
-		replaceExisting = parseResult.mark == 1;
+		replaceExisting = parseResult.hasTag("replacing");
+		particles = !parseResult.hasTag("particles");
+		ambient = parseResult.hasTag("ambient");
+		icon = !parseResult.hasTag("icon");
+		infinite = matchedPattern == 1;
 		if (potionEffect) {
-			potionEffects = (Expression<PotionEffect>) exprs[0];
+			effects = (Expression<PotionEffect>) exprs[0];
 			entities = (Expression<LivingEntity>) exprs[1];
-		} else if (apply) {
+		} else if (infinite) {
 			potions = (Expression<PotionEffectType>) exprs[0];
 			tier = (Expression<Number>) exprs[1];
 			entities = (Expression<LivingEntity>) exprs[2];
-			duration = (Expression<Timespan>) exprs[3];
 		} else {
 			potions = (Expression<PotionEffectType>) exprs[0];
-			entities = (Expression<LivingEntity>) exprs[1];
-		}
-
-		// Ambience and particles
-		switch (matchedPattern) {
-			case 1:
-				ambient = false;
-				particles = true;
-				break;
-			case 2:
-				ambient = true;
-				particles = true;
-				break;
-			case 3:
-				ambient = false;
-				particles = false;
-				break;
+			tier = (Expression<Number>) exprs[1];
+			entities = (Expression<LivingEntity>) exprs[2];
+			duration = (Expression<Timespan>) exprs[3];
 		}
-
 		return true;
 	}
 
 	@Override
-	protected void execute(final Event e) {
+	protected void execute(Event event) {
 		if (potionEffect) {
-			for (LivingEntity livingEntity : entities.getArray(e)) {
-				PotionEffect[] potionEffects = this.potionEffects.getArray(e);
-				PotionEffectUtils.addEffects(livingEntity, potionEffects);
-			}
+			for (LivingEntity livingEntity : entities.getArray(event))
+				PotionEffectUtils.addEffects(livingEntity, effects.getArray(event));
 		} else {
-			final PotionEffectType[] ts = potions.getArray(e);
-			if (ts.length == 0)
-				return;
-			if (!apply) {
-				for (LivingEntity en : entities.getArray(e)) {
-					for (final PotionEffectType t : ts)
-						en.removePotionEffect(t);
-				}
+			PotionEffectType[] potionEffectTypes = potions.getArray(event);
+			if (potionEffectTypes.length == 0)
 				return;
-			}
-			int a = 0;
-			if (tier != null) {
-				final Number amp = tier.getSingle(e);
-				if (amp == null)
-					return;
-				a = amp.intValue() - 1;
-			}
-			int d = DEFAULT_DURATION;
-			if (duration != null) {
-				final Timespan dur = duration.getSingle(e);
-				if (dur == null)
+			int tier = 0;
+			if (this.tier != null)
+				tier = this.tier.getOptionalSingle(event).orElse(1).intValue() - 1;
+
+			int duration = infinite ? (Skript.isRunningMinecraft(1,  19, 4) ? -1 : Integer.MAX_VALUE) : DEFAULT_DURATION;
```

Comment:
`Skript.isRunningMinecraft(1, 19, 4)` check should be initialized outside this method for better performance
```suggestion
			int duration = infinite ? (isRunning1194 ? -1 : Integer.MAX_VALUE) : DEFAULT_DURATION;
```

Reply:
Fair point, but it'll just be comparing against a String. It's not like the classExista method that you're probably thinking of.

Reply:
I think it's worth implementing as a static boolean field